File: cpp/05_049.cpp
std::vector<int> inorder_traversal(Node* root) {
    if (!root) {
        return {};
    }
    
    std::vector<int> result;
    // Left, Root, Right
    std::vector<int> left = inorder_traversal(root->left);
    result.insert(result.end(), left.begin(), left.end());
    
    result.push_back(root->value);
    
    std::vector<int> right = inorder_traversal(root->right);
    result.insert(result.end(), right.begin(), right.end());
    
    return result;
}

File: data structures and algorithms/java/05_049.java
import java.util.*;
import java.util.stream.*;
import java.io.*;
import java.math.*;

public class Main {
    public static void main(String[] args) {
        // Main method code goes here
    }
    
    public class Node {
        public int val;
        public Node left;
        public Node right;
        
        public Node(int val) {
            this.val = val;
        }
    }
    
    public boolean isBalanced(Node root) {
        return checkHeight(root) != -1;
    }
    
    private int checkHeight(Node node) {
        if (node == null) {
            return 0;
        }
        
        int left = checkHeight(node.left);
        if (left == -1) {
            return -1;
        }
            
        int right = checkHeight(node.right);
        if (right == -1) {
            return -1;
        }
            
        if (Math.abs(left - right) > 1) {
            return -1;
        }
            
        return 1 + Math.max(left, right);
    }
}